sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala

```diff
@@ -418,6 +418,63 @@ class CudfLastExcludeNulls(ref: Expression) extends CudfFirstLastBase(ref) {
   override val offset: Int = -1
 }
 
+/**
+ * This class is only used by the M2 class aggregates, do not confuse this with GpuAverage.
+ * In the future, this aggregate class should be removed and the mean values should be
+ * generated in the output of libcudf's M2 aggregate.
+ */
+class CudfMean(ref: Expression) extends CudfAggregate(ref) {
+  @transient val rapidsAvgType: DType = GpuColumnVector.getNonNestedRapidsType(dataType)
+
+  override lazy val updateReductionAggregateInternal: cudf.ColumnVector => cudf.Scalar =
+    throw new UnsupportedOperationException("CudfMean is not supported in reduction")
+  override lazy val mergeReductionAggregateInternal: cudf.ColumnVector => cudf.Scalar =
+    throw new UnsupportedOperationException("CudfMean is not supported in reduction")
+
+  override lazy val updateAggregate: GroupByAggregationOnColumn =
+    GroupByAggregation.mean().onColumn(getOrdinal(ref))
+  override lazy val mergeAggregate: GroupByAggregationOnColumn =
+    GroupByAggregation.mean().onColumn(getOrdinal(ref))
+
+  override def toString(): String = "CudfMeanForM2"
+}
+
+class CudfM2(ref: Expression) extends CudfAggregate(ref) {
+  override lazy val updateReductionAggregateInternal: cudf.ColumnVector => cudf.Scalar =
+    throw new UnsupportedOperationException("M2 aggregation is not yet supported in reduction")
+  override lazy val mergeReductionAggregateInternal: cudf.ColumnVector => cudf.Scalar =
+    throw new UnsupportedOperationException("M2 aggregation is not yet supported in reduction")
+
+  override lazy val updateAggregate: GroupByAggregationOnColumn =
+    GroupByAggregation.M2().onColumn(getOrdinal(ref))
+  override lazy val mergeAggregate: GroupByAggregationOnColumn =
+    GroupByAggregation.mergeM2().onColumn(getOrdinal(ref))
+
+  override def toString(): String = "CudfM2"
+  override def dataType: DataType = DoubleType
+  override def nullable: Boolean = true // TODO: Check nullable
```

Comment:
It is false.  Technically for all of these they are not nullable.

https://github.com/apache/spark/blob/adbea252db0f61ed3aaa45ea20b704bad6c47408/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala#L60-L62

Reply:
Then what happens if the ouput from libcudf is nullable? For example, libcudf M2 outputs null if the input column contains all nulls.

Reply:
Then this is a difference between spark and cudf, that we need to understand to make things work properly. 

Spark starts off with 0.0 as the value for everything.

https://github.com/apache/spark/blob/adbea252db0f61ed3aaa45ea20b704bad6c47408/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala#L77

And then modifies it when it sees a non-null values.

Reply:
So that means we either need to test null outputs from cudf, or replace them with 0.0

Reply:
Could use `GpuCoalesce` to accomplish this, like: https://github.com/NVIDIA/spark-rapids/blob/branch-21.10/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala#L754. 

This can happen in postUpdate / postMerge.